src/merkle/tiered_smt/mod.rs

```diff
@@ -166,6 +166,32 @@ impl TieredSmt {
         TieredSmtProof::new(path, entries).expect("Bug detected, TSMT produced invalid proof")
     }
 
+    /// Returns 
+    /// 1. a partial Merkle tree pre-populated with paths leading up to nodes to be inserted,
+    /// one per key in `keys_to_insert`.
+    /// 2. a map `leaf_hash -> (key, value)` which stores the key/value for keys that already contain a value
+    pub fn get_preinsert_partial_merkle_tree<'a>(
+        &'a self,
+        keys_to_insert: impl IntoIterator<Item = &'a RpoDigest>,
+    ) -> Result<(PartialMerkleTree, BTreeMap<RpoDigest, (RpoDigest, Word)>), MerkleError> {
```

Comment:
A couple of small comments/questions:
- I'd probably name this `get_insert_witness` or something like that.
- I'd probably create a dedicated struct for the return type. Maybe something like `TieredSmtInsertWitness`.
- Why do we need a lifetime here?

Reply:
> Why do we need a lifetime here?

Otherwise surprisingly Rust yells at you: 

> anonymous lifetimes in `impl Trait` are unstable
expected named lifetime parameter